### PR TITLE
Added an overloaded version of toHost() and toAccelerator() that takes a tensor shape, to be used for XLA/TPU compilation

### PR DESCRIPTION
**Pull request by mhong**

### CODE REVIEW

**Comment by mhong** [comment]
@swift-ci please test tensorflow

File: stdlib/public/TensorFlow/Tensor.swift [hunk 1]
```diff
@@ -127,14 +127,45 @@ func _TFHoistable<Scalar>(_ fn: () -> TensorHandle<Scalar>)
...
+  /// Mark memory transfer to host.
+  /// - Parameters:
+  ///   - shape: When sending the tensor to a TF XLA device (including TPU),
+  ///   must specify the tensor shape as required by XLA compilation.
+  @inlinable @inline(__always)
+  func toHost(shape: TensorShape) -> Tensor {
+    // If the `self` tensor resides on TPU, the shape is specified on that
+    // device first, before outfeeding the tensor to CPU, a required step for
+    // sending the tensor to the host.
+    let tensor : Tensor<Scalar> = #tfop("Identity", self, __shapes: [shape])
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Tensor.swift]: Remove space before `:`.

File: stdlib/public/TensorFlow/Tensor.swift [hunk 2]
```diff
@@ -127,14 +127,45 @@ func _TFHoistable<Scalar>(_ fn: () -> TensorHandle<Scalar>)
...
+    return Tensor(handle: _TFReceive(tensor.handle))
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Tensor.swift]: Just `tensor.toHost()` would do it.

File: stdlib/public/TensorFlow/Tensor.swift [hunk 3]
```diff
@@ -127,14 +127,45 @@ func _TFHoistable<Scalar>(_ fn: () -> TensorHandle<Scalar>)
...
 
-/// TODO: Remove when send/receive semantics gets revisited.
 public extension Tensor {
+  /// Mark memory transfer to accelerator.
+  /// - Parameters:
+  ///   - shape: When sending the tensor to a TF XLA device (including TPU),
+  ///   must specify the tensor shape as required by XLA compilation.
+  @inlinable @inline(__always)
+  func toAccelerator(shape: TensorShape) -> Tensor {
+    let tensor = Tensor(handle: _TFSend(handle))
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Tensor.swift]: Just `let tensor = toAccelerator()`

**Review by rxwei — CHANGES_REQUESTED** [review]

**Review by rxwei — APPROVED** [review]
Nice!

File: lib/SILOptimizer/Mandatory/TFLowerGraph.cpp [hunk 4]
```diff
@@ -1230,13 +1230,13 @@ void TFGraphLowering::addTPUEnqueueOp(BuiltinInst *inst, bool isInfeed,
...
           getUserSourceLocation(inst->getDebugLocation()),
-          "TPU infeed enqueue support enqueuing a single tensor -- did "
+          "TPU infeed enqueue supports enqueuing a single tensor -- did "
           "you specify shape?",
           diag::tfop_invalid_tfop);
     else
       internalError(
           getUserSourceLocation(inst->getDebugLocation()),
-          "TPU outfeed enqueue support enqueuing a single tensor -- did "
+          "TPU outfeed enqueue supports enqueuing a single tensor -- did "
```
Comment (lattner) [inline lib/SILOptimizer/Mandatory/TFLowerGraph.cpp]: Random question, but why does outfeed need a shape?  It has an operand which should have a shape, can't it get it from there?

File: stdlib/public/TensorFlow/Tensor.swift [hunk 5]
```diff
@@ -127,14 +127,45 @@ func _TFHoistable<Scalar>(_ fn: () -> TensorHandle<Scalar>)
...
+  @inlinable @inline(__always)
+  func toHost(shape: TensorShape) -> Tensor {
+    // If the `self` tensor resides on TPU, the shape is specified on that
+    // device first, before outfeeding the tensor to CPU, a required step for
+    // sending the tensor to the host.
+    let tensor: Tensor<Scalar> = #tfop("Identity", self, __shapes: [shape])
+    return tensor.toHost()
+  }
+
+  // TODO: Unify these two versions of toAccelerator() by making `shape` above
```
Comment (lattner) [inline stdlib/public/TensorFlow/Tensor.swift]: I actually like the way you have it with two overloads, I don't see a need to merge these.

**Review by lattner — APPROVED** [review]
Looks great!

File: lib/SILOptimizer/Mandatory/TFLowerGraph.cpp (hunk 4 above)
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFLowerGraph.cpp]: Indeed this is not required for infeed enqueue, and I removed the TFLowerGraph logic for its shapes attr. For outfeed enqueue, there is not even a shapes attr (https://b.corp.google.com/issues/110538524).

File: stdlib/public/TensorFlow/Tensor.swift (hunk 5 above)
Comment (mhong) [reply stdlib/public/TensorFlow/Tensor.swift]: k, removed the TODOs here.

**Comment by mhong** [comment]
Thank you both for the review!

**Pull request merged**